### PR TITLE
Action policy cleanup

### PR DESCRIPTION
**Pull request by mishaschwartz**
## Motivation and Context

This PR updates all controller authorization to be through use of the ActionPolicy Gem. 

## Your Changes

**Description**:

- add policies for all routes
- simplify existing non-route policies
- rename non-route policies to be more descriptive
- use `implicit_authorization_target` methods for all controllers so that we can use resourceless `authorize!` to authenticate all routes
- add I18n failure messages for all policies
- clean up code that reports flash messages if policies fail

**Type of change** (select all that apply):

 
- [x] Refactoring (internal change to codebase, without changing functionality)

## Testing

- tested in UI
- TODO: rewrite/update rspec tests for policies

## Checklist

- [x] I have performed a self-review of my own code.
- [x] I have fixed any Hound bot comments. 
- [x] I have verified that the TravisCI tests have passed. 
- [x] I have reviewed the test coverage changes reported on Coveralls. 
- [x] I have added tests for my changes. 
- [x] I have updated the Changelog.md file. 
- [x] I have described any required documentation changes below. 

### Required documentation changes (if applicable)

Style guide added:

https://github.com/MarkUsProject/Markus/wiki/Developer-Guide--Action-Policy-Style-Guide

### CODE REVIEW

File: app/policies/student_policy.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+class StudentPolicy < UserPolicy
+  default_rule :student?
+  authorize :assignment, :grouping, :submission, optional: true
+
+  def run_tests?
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply app/policies/student_policy.rb:6]: vote to ignore

File: spec/policies/ta_policy_spec.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
+describe TaPolicy do
+  let(:context) { { user: user } }
+  let(:record) { user }
+
+  describe_rule :run_tests? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/ta_policy_spec.rb:5]: ignore

File: spec/policies/result_policy_spec.rb [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    end
+    succeed 'user is a ta' do
+      let(:user) { create(:ta) }
+    end
+    failed 'user is a student' do
+      let(:user) { create(:student) }
+    end
+  end
+
+  describe_rule :review? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/result_policy_spec.rb:123]: ignore

File: spec/policies/result_policy_spec.rb [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    end
+    succeed 'user is a ta' do
+      let(:user) { create(:ta) }
+    end
+    succeed 'user is a student' do
+      let(:user) { create(:student) }
+    end
+  end
+
+  describe_rule :run_tests? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/result_policy_spec.rb:16]: ignore

File: spec/policies/key_pair_policy_spec.rb [hunk 5]
```diff
@@ -1,92 +1,62 @@
...
     end
   end
 
-  describe '#manage?', :keep_memory_repos do
-    let(:policy) { :manage? }
-    shared_examples 'user_variation' do |opts: {}|
-      context 'as an admin' do
-        let(:user) { create :admin }
-        include_examples 'vcs_submit_variation', force_pass: opts[:pass_admin], force_fail: opts[:fail_admin]
+  describe_rule :manage? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/key_pair_policy_spec.rb:26]: ignore

File: spec/policies/grouping_policy_spec.rb [hunk 6]
```diff
@@ -1,133 +1,213 @@
...
+    succeed 'grouping has no extension'
+    failed 'grouping has an extension' do
+      before { create :extension, grouping: grouping }
     end
   end
 
-  describe '#view_file_manager?' do
-    subject { described_class.new(grouping, user: user) }
-    context 'as an admin' do
+  describe_rule :view_file_manager? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/grouping_policy_spec.rb:158]: ignore

File: spec/policies/assignment_policy_spec.rb [hunk 7]
```diff
@@ -1,117 +1,217 @@
...
+      let(:user) { create(:student) }
+    end
+  end
 
-      context 'if enable_test is true' do
-        let(:assignment) { build(:assignment, assignment_properties_attributes: { enable_test: true }) }
-
-        context 'if a test group is not configured' do
-          it { is_expected.not_to pass :run_tests?, because_of: :test_groups_exist? }
+  describe_rule :view_test_options? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/assignment_policy_spec.rb:32]: ignore

File: spec/policies/student_policy_spec.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
+describe StudentPolicy do
+  let(:user) { create :student }
+  let(:context) { { user: user } }
+
+  describe_rule :run_tests? do
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply spec/policies/student_policy_spec.rb:5]: ignore

File: app/controllers/submissions_controller.rb [hunk 9]
```diff
@@ -238,25 +197,28 @@ def run_tests
     # .where.not(current_submission_used: nil) potentially makes find fail with RecordNotFound
     test_runs = groupings.select(&:has_non_empty_submission?).map do |g|
       submission = g.current_submission_used
-      authorize! submission, to: :before_release?
+      unless flash_allowance(:error, allowance_to(:run_tests?, current_user, context: { submission: submission })).value
+        return head 400
```
Comment (david-yz-liu) [inline app/controllers/submissions_controller.rb:201]: Please put this on two separate lines

File: app/models/user.rb [hunk 10]
```diff
@@ -161,7 +161,6 @@ def is_reviewer_for?(assignment, result)
       return false
     end
     pr = prs.find {|p| p.result_id == Integer(result.id)}
-
```
Comment (david-yz-liu) [inline app/models/user.rb:164]: Please revert this change

File: app/policies/application_policy.rb [hunk 11]
```diff
@@ -1,3 +1,4 @@
+# Application policy class
 class ApplicationPolicy < ActionPolicy::Base
   # make :manage? a real catch-all
```
Comment (david-yz-liu) [inline app/policies/application_policy.rb:3]: I'm curious about this comment. Is it saying that by default `index?` and `create?` delegate to `manage?`, and if so, can we achieve this instead by using an alias?

Or is it a TODO saying we should set a global `default_rule :manage?` here? (Which is okay with me too.)

File: app/policies/exam_template_policy.rb [hunk 12]
```diff
@@ -4,6 +4,6 @@ class ExamTemplatePolicy < ApplicationPolicy
 
   # Only admin and authorized grader can manage or modify exam templates
```
Comment (david-yz-liu) [inline app/policies/exam_template_policy.rb:5]: You can remove this comment

File: app/policies/grade_entry_form_policy.rb [hunk 13]
```diff
@@ -2,22 +2,18 @@
 class GradeEntryFormPolicy < ApplicationPolicy
   default_rule :manage?
   alias_rule :grades?, :view_summary?, :update_grade?, :get_mark_columns?,
-             :populate_grades_table?, :download?, :upload?, to: :grading?
+             :populate_grades_table?, :download?, :upload?, to: :grade?
 
   # Only admin and grader can grade the students result.
```
Comment (david-yz-liu) [inline app/policies/grade_entry_form_policy.rb:7]: You can delete this comment

File: spec/controllers/course_summaries_controller_spec.rb [hunk 14]
```diff
@@ -130,7 +130,7 @@
 
     it 'not be able to CSV graders report' do
       get_as @grader, :download_csv_grades_report
-      expect(response.status).to eq(404)
+      expect(response.status).to eq(403)
```
Comment (david-yz-liu) [inline spec/controllers/course_summaries_controller_spec.rb:133]: Since you're changing all of these anyway, do you want to switch to the preferred `expect(response).to have_http_status(403)`?

File: spec/controllers/students_controller_spec.rb [hunk 15]
```diff
@@ -103,8 +103,7 @@
   end
 
   describe 'User is a Student' do
-    include ERB::Util
```
Comment (david-yz-liu) [inline spec/controllers/students_controller_spec.rb:106]: This is good---there's going to be a merge conflict here because I merged in https://github.com/MarkUsProject/Markus/pull/4922/files#diff-5f44c1c580dbc94788e40f1aecf6f711215f30ce917214dc8d25765105cc4a37R15. (FYI)

**Review by david-yz-liu — COMMENTED** [review]
Cool, I just left a few minor style comments, and one comment about a merge conflict. Lots of code change, but everything seems pretty straightforward. I'll look at the wiki now too.

File: app/controllers/submissions_controller.rb (hunk 9 above)
Comment (mishaschwartz) [reply app/controllers/submissions_controller.rb:201]: rubocop will complain...

File: app/policies/application_policy.rb (hunk 11 above)
Comment (mishaschwartz) [reply app/policies/application_policy.rb:3]: It's because index? and create? are defined in the superclass ActionPolicy::Base so setting a default rule won't actually work here (https://actionpolicy.evilmartians.io/#/custom_policy?id=custom-base-policy). 
We can use aliases but that loses some flexibility if we want to show different error messages in the future. But since we're not differentiating right now, I'm happy to move them over to aliases.

File: app/controllers/submissions_controller.rb [hunk 16]
```diff
@@ -238,25 +197,29 @@ def run_tests
     # .where.not(current_submission_used: nil) potentially makes find fail with RecordNotFound
     test_runs = groupings.select(&:has_non_empty_submission?).map do |g|
       submission = g.current_submission_used
-      authorize! submission, to: :before_release?
+      unless flash_allowance(:error, allowance_to(:run_tests?, current_user, context: { submission: submission })).value
+        head 400
+        return
```
[the remark this answers is not in the document]
Comment (mishaschwartz) [reply app/controllers/submissions_controller.rb:202]: vote to ignore given comment above

File: spec/controllers/students_controller_spec.rb (hunk 15 above)
Comment (mishaschwartz) [reply spec/controllers/students_controller_spec.rb]: nice! but it looks like it handled the merge ok

File: app/controllers/submissions_controller.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (david-yz-liu) [reply app/controllers/submissions_controller.rb:202]: Thanks! Yes let's ignore

File: app/policies/application_policy.rb (hunk 11 above)
Comment (david-yz-liu) [reply app/policies/application_policy.rb]: Thanks for the link!
